src/groupByKey.ts

```diff
@@ -2,6 +2,21 @@ type GroupsMap<T> = {
   [key: string]: T[];
 };
 
-export function groupByKey(items, key) {
-  // write code here;
+export function groupByKey<T>(items: T[], key: keyof T): GroupsMap<T> {
+  const groupedData: GroupsMap<T> = {};
+  const allKeysValues: Array<string> = [];
+
+  for (const item of items) {
+    const value = `${item[key]}`;
+
+    if (!allKeysValues.includes(value)) {
+      allKeysValues.push(value);
+    }
```

Comment:
The creation of an array to store all key values is unnecessary and results in extra iterations over the data. This can be optimized by directly checking and assigning to the groupedData object within a single loop.